Sources/Profile-Export.php

```diff
@@ -0,0 +1,1956 @@
+			),
+			'txt_posts_heading' => array(
+				'value' => $txt['posts'],
+			),
+			'txt_personal_messages_heading' => array(
+				'value' => $txt['personal_messages'],
+			),
+			'txt_view_source_button' => array(
+				'value' => $txt['export_view_source_button'],
+			),
+			'txt_download_original' => array(
+				'value' => $txt['export_download_original'],
+			),
+			'txt_help' => array(
+				'value' => $txt['help'],
+			),
+			'txt_terms_rules' => array(
+				'value' => $txt['terms_and_rules'],
+			),
+			'txt_go_up' => array(
+				'value' => $txt['go_up'],
+			),
+			'txt_pages' => array(
+				'value' => $txt['pages'],
+			),
+		);
+
+		// Let mods adjust the XSLT variables.
+		call_integration_hook('integrate_export_xslt_variables', array(&$xslt_variables, $format));
+
+		$idhash = hash_hmac('sha1', $uid, get_auth_secret());
+		$xslt_variables['dltoken'] = array(
+			'value' => hash_hmac('sha1', $idhash, get_auth_secret())
+		);
+
+		// Efficiency = good.
+		$xslt_key = $smcFunc['json_encode'](array($format, $uid, $xslt_variables));
+		if (isset($xslts[$xslt_key]))
+			return $xslts[$xslt_key];
+
+		if ($format == 'XML_XSLT')
+		{
+			$doctype = implode("\n", array(
+				'<!--',
+				"\t" . $txt['export_open_in_browser'],
+				'-->',
+				'<?xml-stylesheet type="text/xsl" href="#stylesheet"?>',
+				'<!DOCTYPE smf:xml-feed [',
+				'<!ATTLIST xsl:stylesheet',
+				'id ID #REQUIRED>',
+				']>',
+			));
+
+			$stylesheet['header'] = "\n" . implode("\n", array(
+				'',
+				"\t" . '<xsl:stylesheet version="1.0" xmlns:xsl="' . $xslt_ns . '" xmlns:html="' . $html_ns . '" xmlns:smf="' . $smf_ns . '" exclude-result-prefixes="smf html" id="stylesheet">',
+				'',
+				"\t\t" . '<xsl:template match="xsl:stylesheet"/>',
+				"\t\t" . '<xsl:template match="xsl:stylesheet" mode="detailedinfo"/>',
+			));
+		}
+		else
+		{
+			$doctype = '';
+			$stylesheet['header'] = implode("\n", array(
+				'<?xml version="1.0" encoding="' . $context['character_set'] . '"?' . '>',
+				'<xsl:stylesheet version="1.0" xmlns:xsl="' . $xslt_ns . '" xmlns:html="' . $html_ns . '" xmlns:smf="' . $smf_ns . '" exclude-result-prefixes="smf html">',
+			));
+		}
+
+		// Output control settings.
+		$stylesheet['output_control'] = '
+		<xsl:output method="html" encoding="utf-8" indent="yes"/>
+		<xsl:strip-space elements="*"/>';
+
+		// Insert the XSLT variables.
+		$stylesheet['variables'] = '';
+
+		foreach ($xslt_variables as $name => $var)
+		{
+			$element = !empty($var['param']) ? 'param' : 'variable';
+
+			$stylesheet['variables'] .= "\n\t\t" . '<xsl:' . $element . ' name="' . $name . '"';
+
+			if (isset($var['xpath']))
+				$stylesheet['variables'] .= ' select="' . $var['value'] . '"/>';
+			else
+				$stylesheet['variables'] .= '>' . (!empty($var['no_cdata_parse']) ? $var['value'] : cdata_parse($var['value'])) . '</xsl:' . $element . '>';
+		}
+
+		// The top-level template. Creates the shell of the HTML document.
+		$stylesheet['html'] = '
```

Comment:
Can these be moved to a template file, perhaps one function for each and returning the string rather than echoing it.

Reply:
That's a good idea.

Reply:
Sounds like changes are coming...  I'll wait until after the new commits to re-test.  Just let me know when they are complete.

Reply:
It's done now, @sbulen: 830c32b

It doesn't change much, really. It mostly just shuffles some existing code around between different files. I also took the opportunity to make this system easily extensible in case we find ourselves needing other XSLT stylesheets in the future.